### PR TITLE
rgw: orphan tool shouldn't clean up head objects

### PR DESCRIPTION
**Pull request by yehudasa**
Fixes: #12958

Head objects are mutable, so removing them can race with object removal
and a later recreation, so we might end up cleaning them up when we don't
need to.

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>

### CODE REVIEW

**Comment by dachary** [comment]
please ignore the bot false negative ( http://tracker.ceph.com/issues/13176 )

File: src/rgw/rgw_orphan.cc [hunk 1]
```diff
@@ -277,9 +277,28 @@ int RGWOrphanSearch::build_all_oids_index()
...
-      name += " (@" + locator + ")";  
+    ssize_t pos = oid.find('_');
+    if (pos < 0) {
+      cout << "unidentified oid: " << oid << ", skipping" << std::endl;
+      /* what is this object, oids should be in the format of <bucket marker>_<obj>,
+       * skip this entry
+       */
+      continue;
+    }
+    string stripped_oid = oid.substr(pos + 1);
```
Comment (cbodley) [inline src/rgw/rgw_orphan.cc]: I'm concerned about the +1 here to skip over the '\_'. The first check in parse_raw_oid() is
```c++
    if (oid[0] != '_') {
      *obj_name = oid;
      return true;
    }
```
So it looks to me like stripping "\<bucket\>\_" from the beginning will prevent parse_raw_oid() from ever parsing a namespace, whereas leaving that trailing '\_' should work as intended.

Comment (yehudasa) [inline src/rgw/rgw_orphan.cc]: @cbodley the "<bucket>_" is fixed for any oid, whether it has a namespace or whether it doesn't have it. The next char after it is the one that defines whether there's going to be a namespace or not. So an oid that has a namespace will have two underscores, not one.

Comment (cbodley) [inline src/rgw/rgw_orphan.cc]: @yehudasa ok thanks! then the fix looks good to me

**Comment by cbodley** [comment]
Retargeted against infernalis and merged manually.

**Pull request closed without merging**